lib/services/unified_resource.go

```diff
@@ -0,0 +1,464 @@
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err, "setting defaults for unified resource watcher config")
+	}
+
+	if _, err := newResourceWatcher(ctx, resourceCache, cfg); err != nil {
+		return trace.Wrap(err, "creating a new unified resource watcher")
+	}
+	return nil
+}
+
+func resourceKey(resource types.Resource) []byte {
+	return backend.Key(prefix, resource.GetName(), resource.GetKind())
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := c.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.getAndUpdateDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.getAndUpdateKubes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.getAndUpdateApps(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.getAndUpdateSAMLApps(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.getAndUpdateDesktops(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	c.stale = false
+	c.defineCollectorAsInitialized()
+	return nil
+}
+
+// getAndUpdateNodes will get nodes and update the current tree with each Node
+func (c *UnifiedResourceCache) getAndUpdateNodes(ctx context.Context) error {
+	newNodes, err := c.ResourceGetter.GetNodes(ctx, apidefaults.Namespace)
+	if err != nil {
+		return trace.Wrap(err, "getting nodes for unified resource watcher")
+	}
+
+	return trace.Wrap(putResources[types.Server](c, newNodes))
+}
+
+// getAndUpdateDatabases will get database servers and update the current tree with each DatabaseServer
+func (c *UnifiedResourceCache) getAndUpdateDatabases(ctx context.Context) error {
+	newDbs, err := c.GetDatabaseServers(ctx, apidefaults.Namespace)
+	if err != nil {
+		return trace.Wrap(err, "getting databases for unified resource watcher")
+	}
+
+	return trace.Wrap(putResources[types.DatabaseServer](c, newDbs))
+}
+
+func (c *UnifiedResourceCache) refreshStaleResources(ctx context.Context) error {
+	c.mu.Lock()
+	if !c.stale && c.isInitialized() {
+		c.mu.Unlock()
+		return nil
+	}
+	c.mu.Unlock()
+
+	_, err := utils.FnCacheGet(ctx, c.cache, "unified_resources", func(ctx context.Context) (any, error) {
+		fallbackCache := &UnifiedResourceCache{
+			cfg: c.cfg,
+			tree: btree.NewG(c.cfg.BTreeDegree, func(a, b *item) bool {
+				return a.Less(b)
+			}),
+			ResourceGetter:  c.ResourceGetter,
+			initializationC: make(chan struct{}),
+		}
+		if err := fallbackCache.getResourcesAndUpdateCurrent(ctx); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		// There is a chance that the watcher reinitialized while
+		// getting resources happened above. Check if we are still stale
+		// now that the lock is held to ensure that the refresh is
+		// still necessary.
+		if !c.stale {
+			return nil, nil
+		}
+
+		c.tree = fallbackCache.tree
```

Comment:
Having the ttl-based fallback tree take up the same slot as the primary tree complicates the concurrency model, since this means it isn't safe to read from the fallback tree while the primary tree is being initialized (since you might end up reading from a partially-initialized state). This isn't that crazy of a thing to workaround, the main thing is to make sure that your initialization/reset logic holds the lock continuously, so that no reads happen between clearing the ttl-tree state and completion of all inserts of new values.

Not a requirement, but I'm a fan of creating helper methods that handle deciding which underlying state to access and abstract over the different locking requirements of the two access patterns.  We did a more complex version of this in `lib/cache` with the "read guard", but I think a simpler model of just passing a closure into a helper function would work pretty well here:

```golang
// read applies the supplied closure to either the primary tree or the ttl-based fallback tree depending on
// wether or not the cache is currently healthy.  locking is handled internally and the passed-in tree should
// not be accessed after the closure completes.
func (c *UnifiedResourceCache) read(fn func(tree *btree.BTreeG[*item])) error {
    c.mu.Lock()
    
    if !c.stale {
        fn(c.tree)
        c.mu.Unlock()
        return nil
    }
    
    c.mu.Unlock()
    ttlTree, err := utils.FnCacheGet(ctx, c.cache, "unified_resources", func(ctx context.Context) (*btree.BTreeG[*item], error) {
        // ...
        return fallbackCache.tree
    })
    c.mu.Lock()

    if !c.stale {
        // primary became healthy while we were waiting
        fn(c.tree)
        c.mu.Unlock()
        return nil
    }
    c.mu.Unlock()
    
    if err != nil {
        // ttl-tree setup failed
        return trace.Wrap(err)
    }
    
    fn(ttlTree)
    return nil
}


func (c *UnifiedResourceCache) example() error {
    err := c.read(func(tree *btree.BTreeG[*item]) {
        // do stuff with the tree here while the lock is held
    })
    return trace.Wrap(err)
}
```